crates/sl-paillier/src/lib.rs

```diff
@@ -4,46 +4,173 @@
 use std::ops::Deref;
 
 use crypto_bigint::modular::runtime_mod::{DynResidue, DynResidueParams};
-use crypto_bigint::{Encoding, NonZero, RandomMod, Split, Uint};
+use crypto_bigint::{Bounded, Encoding, NonZero, RandomMod, Split, Uint};
 use crypto_bigint::{U1024, U2048, U4096};
 
 use crypto_primes::generate_prime_with_rng;
 
 use rand_core::CryptoRngCore;
 
-// #[cfg(test)]
-// #[macro_use(quickcheck)]
-// extern crate quickcheck_macros;
+// print-type-size type: `SK<64, 32, 16>`: 5400 bytes, alignment: 8 bytes
+// print-type-size     field `.phi`: 256 bytes
+// print-type-size     field `.inv_phi`: 256 bytes
+// print-type-size     field `.p`: 128 bytes
+// print-type-size     field `.hp`: 128 bytes
+// print-type-size     field `.q`: 128 bytes
+// print-type-size     field `.hq`: 128 bytes
+// print-type-size     field `.pk`: 2312 bytes
+// print-type-size     field `.pp_params`: 1032 bytes
+// print-type-size     field `.qq_params`: 1032 bytes
+
+pub type SK2048 = SK<{ U4096::LIMBS }, { U2048::LIMBS }, { U1024::LIMBS }>;
+pub type PK2048 = PK<{ U4096::LIMBS }, { U2048::LIMBS }>;
+pub type MinimalSK2048 = MinimalSK<{ U1024::LIMBS }>;
+pub type MinimalPK2048 = MinimalPK<{ U4096::LIMBS }, { U2048::LIMBS }>;
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct MinimalSK<const P: usize>
+where
+    Uint<P>: Bounded + Encoding,
+{
+    pub p: Uint<P>,
+    pub q: Uint<P>,
+}
 
-pub mod paillier {
+#[derive(Debug, Clone, Copy)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct MinimalPK<const NN: usize, const N: usize>
+where
+    Uint<N>: Bounded + Encoding,
+    Uint<NN>: Bounded + Encoding,
+{
+    pub n: NonZero<Uint<N>>,
+    pub nn: Uint<NN>,
```

Comment:
`nn = n*n`

Do we have to  serialize it?

Reply:
We use this value while verifying some proofs. But after check it's okay to remove. Removed in latest commit.